### ISSUE TITLE
Failure to jump to the character location on error when processing past the error

### ISSUE THREAD

**Issue by JasonGross** [psteckler/ProofGeneral#26]
In
```coq
Definition x := 1.
Section foo.
  Inductive x := a.
End foo.
Section bar.
  Context (y : nat).
  Goal True.
    constructor.
  Qed.
End bar.
```
If I go to the end and do `proof-goto-point`, then the cursor stays at the end, and I get no indication that there's an error.  (The line gets bolded; it should get colored red, and the cursor should jump to that point.  Or, at least, there should be a setting that makes the cursor jump to that point.  Async is useful for proofs, but is counter-productive for definitions if it hides the fact that I'm editing.)

**Comment by psteckler** [comment]
I'm seeing different behavior. The line "Inductive x := a." is turned red, and the cursor is at the end of the preceding line. FWIW, that's what CoqIDE does.

**Comment by JasonGross** [comment]
Relevant details: I'm using emacs 24 in screen over ssh, and have `proof-prog-name-ask` set to `t`.

If I paste the above in the buffer, and then type `M-> M-: (proof-goto-point) RET RET`, my cursor remains at the end of the document.

![bad-pg](https://cloud.githubusercontent.com/assets/396076/18647751/5e2b73ee-7e6c-11e6-917e-370fc65d96f7.gif)

**Comment by psteckler** [comment]
I can reproduce the issue running this script over an ssh connection.

**Comment by psteckler** [comment]
Hmmm, I see the cursor-stays-put behavior with my somewhat modified v8.6 branch, even with an X emacs client, but it's close to OK with v8.5pl2.

In 8.5, there's an XML feedback message that indicates that the the Inductive line should be colored red, and a fail-value that says control should go back to the state preceding that line.

In 8.6, I get  the feedback message, but not the fail-value. I suspect you're using 8.6, then.

Can you attach the *coq-log* that you see?

The 8.6 behavior may be a bug in Coq.

**Comment by JasonGross** [comment]
Yes, this is 8.6 from a day or two ago.


*coq-log*
```
*emacs:
<call val="Init"><option val="none"/></call>

*coqtop:
<value val="good"><state_id val="2"/></value>
*emacs:
<call val="Query"><pair><string>Remove Search Blacklist "Private_" "_subproof".
Add Search Blacklist "Private_" "_subproof".</string>
<state_id val="2"/></pair>
</call>

*coqtop:
<feedback object="state" route="0"><state_id val="2"/><feedback_content val="processed"/></feedback>
*coqtop:
<value val="good"><string></string></value>
*emacs:
<call val="SetOptions"><list><pair><list><string>Printing</string>
<string>Depth</string>
</list>
<option_value val="intvalue"><option val="some"><int>50</int>
</option>
</option_value>
</pair>
</list>
</call>

*coqtop:
<value val="good"><unit/></value>
*emacs:
<call val="SetOptions"><list><pair><list><string>Printing</string>
<string>Width</string>
</list>
<option_value val="intvalue"><option val="some"><int>133</int>
</option>
</option_value>
</pair>
</list>
</call>

*coqtop:
<value val="good"><unit/></value>
*emacs:
<call val="Add"><pair><pair><string>Definition&nbsp;x&nbsp;:=&nbsp;1.</string>
<int>-1</int>
</pair>
<pair><state_id val="2"/><bool val="true"/></pair>
</pair>
</call>

*coqtop:
<value val="good"><pair><state_id val="4"/><pair><union val="in_l"><unit/></union><string></string></pair></pair></value>
*emacs:
<call val="Add"><pair><pair><string>Section&nbsp;foo.</string>
<int>-2</int>
</pair>
<pair><state_id val="4"/><bool val="true"/></pair>
</pair>
</call>

*coqtop:
<value val="good"><pair><state_id val="5"/><pair><union val="in_l"><unit/></union><string></string></pair></pair></value>
*emacs:
<call val="Add"><pair><pair><string>Inductive&nbsp;x&nbsp;:=&nbsp;a.</string>
<int>-3</int>
</pair>
<pair><state_id val="5"/><bool val="true"/></pair>
</pair>
</call>

*coqtop:
<value val="good"><pair><state_id val="6"/><pair><union val="in_l"><unit/></union><string></string></pair></pair></value>
*emacs:
<call val="Add"><pair><pair><string>End&nbsp;foo.</string>
<int>-4</int>
</pair>
<pair><state_id val="6"/><bool val="true"/></pair>
</pair>
</call>

*coqtop:
<feedback object="state" route="0"><state_id val="7"/><feedback_content val="processingin"><string>master</string></feedback_content></feedback><feedback object="state" route="0"><state_id val="6"/><feedback_content val="processingin"><string>master</string></feedback_content></feedback><feedback object="state" route="0"><state_id val="5"/><feedback_content val="processingin"><string>master</string></feedback_content></feedback><feedback object="state" route="0"><state_id val="4"/><feedback_content val="processingin"><string>master</string></feedback_content></feedback><feedback object="state" route="0"><state_id val="2"/><feedback_content val="processed"/></feedback><message><message_level val="info"/><option val="none"/><richpp><_>x&nbsp;is&nbsp;defined</_></richpp></message><feedback object="state" route="0"><state_id val="4"/><feedback_content val="processed"/></feedback><feedback object="state" route="0"><state_id val="5"/><feedback_content val="processed"/></feedback><feedback object="state" route="0"><state_id val="6"/><feedback_content val="message"><message><message_level val="error"/><option val="some"><loc start="0" stop="17"/></option><richpp><_><message.error>Error:</message.error>&nbsp;x&nbsp;already&nbsp;exists.</_></richpp></message></feedback_content></feedback><feedback object="state" route="0"><state_id val="7"/><feedback_content val="processed"/></feedback><value val="good"><pair><state_id val="7"/><pair><union val="in_l"><unit/></union><string></string></pair></pair></value>
*emacs:
<call val="Add"><pair><pair><string>Section&nbsp;bar.</string>
<int>-5</int>
</pair>
<pair><state_id val="7"/><bool val="true"/></pair>
</pair>
</call>

*coqtop:
<value val="good"><pair><state_id val="8"/><pair><union val="in_l"><unit/></union><string></string></pair></pair></value>
*emacs:
<call val="Add"><pair><pair><string>Context&nbsp;(y&nbsp;:&nbsp;nat).</string>
<int>-6</int>
</pair>
<pair><state_id val="8"/><bool val="true"/></pair>
</pair>
</call>

*coqtop:
<feedback object="state" route="0"><state_id val="8"/><feedback_content val="processingin"><string>master</string></feedback_content></feedback><feedback object="state" route="0"><state_id val="7"/><feedback_content val="processed"/></feedback><feedback object="state" route="0"><state_id val="8"/><feedback_content val="processed"/></feedback><feedback object="state" route="0"><state_id val="9"/><feedback_content val="processingin"><string>master</string></feedback_content></feedback><feedback object="state" route="0"><state_id val="8"/><feedback_content val="processed"/></feedback><message><message_level val="info"/><option val="none"/><richpp><_>y&nbsp;is&nbsp;declared</_></richpp></message><feedback object="state" route="0"><state_id val="9"/><feedback_content val="processed"/></feedback><value val="good"><pair><state_id val="9"/><pair><union val="in_l"><unit/></union><string></string></pair></pair></value>
*emacs:
<call val="Add"><pair><pair><string>Goal&nbsp;True.</string>
<int>-7</int>
</pair>
<pair><state_id val="9"/><bool val="true"/></pair>
</pair>
</call>

*coqtop:
<value val="good"><pair><state_id val="10"/><pair><union val="in_l"><unit/></union><string></string></pair></pair></value>
*emacs:
<call val="Add"><pair><pair><string>constructor.</string>
<int>-8</int>
</pair>
<pair><state_id val="10"/><bool val="true"/></pair>
</pair>
</call>

*coqtop:
<value val="good"><pair><state_id val="11"/><pair><union val="in_l"><unit/></union><string></string></pair></pair></value>
*emacs:
<call val="Add"><pair><pair><string>Qed.</string>
<int>-9</int>
</pair>
<pair><state_id val="11"/><bool val="true"/></pair>
</pair>
</call>

*coqtop:
<value val="good"><pair><state_id val="12"/><pair><union val="in_l"><unit/></union><string></string></pair></pair></value>
*emacs:
<call val="Add"><pair><pair><string>End&nbsp;bar.</string>
<int>-10</int>
</pair>
<pair><state_id val="12"/><bool val="true"/></pair>
</pair>
</call>

*coqtop:
<feedback object="state" route="0"><state_id val="13"/><feedback_content val="processingin"><string>master</string></feedback_content></feedback>
*coqtop:
<feedback object="state" route="0"><state_id val="12"/><feedback_content val="processingin"><string>master</string></feedback_content></feedback>
*coqtop:
<feedback object="state" route="0"><state_id val="11"/><feedback_content val="processingin"><string>master</string></feedback_content></feedback>
*coqtop:
<feedback object="state" route="0"><state_id val="10"/><feedback_content val="processingin"><string>master</string></feedback_content></feedback>
*coqtop:
<feedback object="state" route="0"><state_id val="9"/><feedback_content val="processed"/></feedback>
*coqtop:
<feedback object="state" route="0"><state_id val="10"/><feedback_content val="processed"/></feedback>
*coqtop:
<feedback object="state" route="0"><state_id val="11"/><feedback_content val="processed"/></feedback>
*coqtop:
<feedback object="state" route="0"><state_id val="11"/><feedback_content val="processed"/></feedback>
*coqtop:
<feedback object="state" route="0"><state_id val="10"/><feedback_content val="processed"/></feedback>
*coqtop:
<message><message_level val="info"/><option val="none"/><richpp><_>Unnamed_thm&nbsp;is&nbsp;defined</_></richpp></message><feedback object="state" route="0"><state_id val="12"/><feedback_content val="processed"/></feedback><feedback object="state" route="0"><state_id val="13"/><feedback_content val="processed"/></feedback><value val="good"><pair><state_id val="13"/><pair><union val="in_l"><unit/></union><string></string></pair></pair></value>
*emacs:
<call val="Goal"><unit/></call>

*coqtop:
<feedback object="state" route="0"><state_id val="13"/><feedback_content val="processed"/></feedback><value val="good"><option val="none"/></value>
*emacs:
<call val="Status"><bool val="false"/></call>

*coqtop:
<feedback object="state" route="0"><state_id val="13"/><feedback_content val="processed"/></feedback><value val="good"><status><list><string>Top</string></list><option val="none"/><list/><int>0</int></status></value>
```

**Comment by psteckler** [comment]
Thanks, that confirms what I see here. The XML in 8.6  does not provide the information to position the cursor at the error location. The same behavior is observable with CoqIDE 8.6.

I've filed bug 5092 in the Coq Bugzilla.

**Comment by psteckler** [comment]
The resolution of bug 5092 was WONTFIX.

The error message does contain the state id where the error occurred, so it could, in theory, be used to trigger a jump back to that state. But in the case that a proof is being computed by a worker, you'd not be able to process later proofs. I suppose that's why the change was made.

**Comment by JasonGross** [comment]
There are roughly three modes of development that I want, from most frequent to least frequent:
1. I'm not sure if this code has errors; please process it (synchronously, or asynchronously if possible) and jump to the first error you can find as quickly as possible.  (Importantly, I should be able to keep typing while processing is going on, and add things to the to-be-processed buffer.  It'd be nice for proofs to be processed in the background / in parallel, but it's not necessary.)
2. I'm pretty sure this code does not have errors.  Get me to the end as quickly as possible and let me do development while proofs are running in the background.  Do catch my attention in some way (not by moving the cursor) if there are errors.
3. This code probably has errors.  But I want to keep working on the end anyway; please get me there quickly and don't interrupt me when you find errors.  (But do give some indication, like scroll bar and background colors.)

I'd like a variant of `proof-goto-point` for each of these, so I can have a keyboard shortcut for each.  (2 and 3 might end up being the same mode.)  Can this be done?

**Comment by psteckler** [comment]
There's a distinction between cursor placement and point of control. The 8.5 behavior was to suggest to IDEs to request a return of control to the point of an error. In 8.6, that's gone. 

We could implement (1) by returning control based on error messages, as I mentioned in an earlier comment. That kind of violates what coqtop is tell the IDE to do, but OK. If the point of control hasn't changed, you probably don't want automatic cursor warping, too jarring.

 For (2) and (3), there could be a keybinding or other mechanism to cycle among errors, along with notification of errors (maybe via text properties on the Emacs mode line).

**Comment by gares** [comment]
On Wed, Sep 21, 2016 at 09:42:31AM -0700, psteckler wrote:
> There's a distinction between cursor placement and point of control.
> The 8.5 behavior was to suggest to IDEs to request a return of control
> to the point of an error. In 8.6, that's gone. 

I'd put it like that.

In 8.5:
- there were errors requesting the UI to stop processing
- and errors (inside proofs) not requesting that

In 8.6:
- there are fewer errors requesting the UI to stop processing
- and more errors not requesting that

In both cases you have to handle both behaviors, only Coq knows/decides
in which class the error is.

E.g. if in 8.6 you jump immediately after a broken Definition/Inductive,
you get an error of the first class.  If you jump after a working
sentence that occurs after the erroneous one, then you get an error of
the second class.  The only invariant relevant to the UI is that the tip
is a valid state.

All Jason needs is a visual feedback of errors that donesn't require him
to scroll the buffer up an down just for that.  This covers all his use
cases (it just changes what he does when he is notified about the
errors: react immediately or leave them on the side).

CoqIDE, Coqoon, Jedit, Eclipse for Java/Scala, ...  provide a gutter
(i.e. red spots in a scroll bar like widget) to signal errors as well as
an error summary (a window with a list of all error messages).

Both things are visible in the screenshot attached to PR 173 in case
you don't see what I mean (the gutter is horizontal in CoqIDE, courtesy of PMP).

How to code any of these widgets, or an alternative one in PG, is up
to you Paul.  Early versions of CoqIDE just had just a red number (of
errors) in the left right corner, not perfect but way better than scrolling.

Best
-- 
Enrico Tassi

**Comment by gares** [comment]
Let me just add that clicking on an error message in the error summary (or in the gutter) jumps to the error, so that "moving the cursor automatically" is (almost) never necessary.  In this way, use case 1 and 2 can be quickly achieved by the user clicking on the error message.

**Comment by gares** [comment]
for reference https://github.com/coq/coq/pull/173

**Comment by JasonGross** [comment]
> clicking on an error message in the error summary (or in the gutter) jumps to the error

Note that unlike CoqIDE, emacs also has a terminal GUI, which I frequently use over SSH; in this mode, clicking is not an option, and how much I have to do to get to the error location really matters in terms of ease of use.

**Comment by gares** [comment]
I imagine emacs has a standard window to report errors (when you run make for example).  Paul could just reuse that "widget" to list errors.  Dunno how to jump to the error location in such cases, but I guess it works over SSH too...

BTW, if I start vim via SSH it reacts to my mouse clicks...

**Comment by JasonGross** [comment]
> BTW, if I start vim via SSH it reacts to my mouse clicks...

Oh, er, maybe it's because I'm using emacs in screen over mosh+ssh..

**Comment by psteckler** [comment]
OK, there's now a cheapo-gutter-widget in PG, to allow viewing error locations.

Can we close this issue?
